System.Text.Json source generator fails if [Required(ErrorMessage = "...")] attribute is present.

### Description

I tried to use the Json source generator with a class which has a [Required] attribute with an error message, but the generator succeeds only when the ErrorMessage is not specified. I tried it with other Validation attributes (StringLength, Range), and I got the same result.

### Reproduction Steps

Add the following two classes in a .NET 6.0 project, then try to build it.

```csharp
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

public class TestDto
{
    [Required(ErrorMessage = "asdf")]
    public string Name { get; set; }
}

[JsonSerializable(typeof(TestDto))]
public partial class Serializers : JsonSerializerContext
{

}

```

### Expected behavior

The source generator should generate the properties and the build should succeed.

### Actual behavior

The source generator and the build fails and I get an error in the Output window:

```
Build started...
1>------ Build started: Project: SystemTextJsonTest, Configuration: Debug Any CPU ------
1>CSC : warning CS8785: Generator 'JsonSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'InvalidOperationException' with message 'Sequence contains no elements'
1>D:\src\SystemTextJsonTest\SystemTextJsonTest\TestDto.cs(11,22,11,33): error CS0534: 'Serializers' does not implement inherited abstract member 'JsonSerializerContext.GetTypeInfo(Type)'
1>D:\src\SystemTextJsonTest\SystemTextJsonTest\TestDto.cs(11,22,11,33): error CS0534: 'Serializers' does not implement inherited abstract member 'JsonSerializerContext.GeneratedSerializerOptions.get'
1>Done building project "SystemTextJsonTest.csproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

### Regression?

_No response_

### Known Workarounds

Remove the ErrorMessage, ErrorMessageResourceName and ErrorMessageResourceType properties from the validation attributes.

### Configuration

.NET 6.0.100
Microsoft Windows [Version 10.0.19043.1348]
x64

### Other information

_No response_

--- a/src/libraries/System.Text.Json/gen/Reflection/CustomAttributeDataWrapper.cs
+++ b/src/libraries/System.Text.Json/gen/Reflection/CustomAttributeDataWrapper.cs
@@ -15,7 +15,7 @@ namespace System.Text.Json.Reflection
             var namedArguments = new List<CustomAttributeNamedArgument>();
             foreach (KeyValuePair<string, TypedConstant> na in a.NamedArguments)
             {
-                var member = a.AttributeClass!.GetMembers(na.Key).First();
+                var member = a.AttributeClass.BaseTypes().SelectMany(t => t.GetMembers(na.Key)).First();
 
                 MemberInfo memberInfo = member is IPropertySymbol
                     ? new PropertyInfoWrapper((IPropertySymbol)member, metadataLoadContext)

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/ContextClasses.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/ContextClasses.cs
@@ -47,6 +47,8 @@ namespace System.Text.Json.SourceGeneration.Tests
         public JsonTypeInfo<ClassWithBadCustomConverter> ClassWithBadCustomConverter { get; }
         public JsonTypeInfo<StructWithBadCustomConverter> StructWithBadCustomConverter { get; }
         public JsonTypeInfo<PersonStruct?> NullablePersonStruct { get; }
+        public JsonTypeInfo<TypeWithValidationAttributes> TypeWithValidationAttributes { get; }
+        public JsonTypeInfo<TypeWithDerivedAttribute> TypeWithDerivedAttribute { get; }
     }
 
     internal partial class JsonContext : JsonSerializerContext

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/MetadataAndSerializationContextTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/MetadataAndSerializationContextTests.cs
@@ -41,6 +41,8 @@ namespace System.Text.Json.SourceGeneration.Tests
     [JsonSerializable(typeof(ClassWithBadCustomConverter))]
     [JsonSerializable(typeof(StructWithBadCustomConverter))]
     [JsonSerializable(typeof(PersonStruct?))]
+    [JsonSerializable(typeof(TypeWithValidationAttributes))]
+    [JsonSerializable(typeof(TypeWithDerivedAttribute))]
     internal partial class MetadataAndSerializationContext : JsonSerializerContext, ITestContext
     {
         public JsonSourceGenerationMode JsonSourceGenerationMode => JsonSourceGenerationMode.Default;
@@ -88,6 +90,8 @@ namespace System.Text.Json.SourceGeneration.Tests
             Assert.Throws<InvalidOperationException>(() => MetadataAndSerializationContext.Default.StructWithBadCustomConverter);
             Assert.Null(MetadataAndSerializationContext.Default.NullablePersonStruct.SerializeHandler);
             Assert.NotNull(MetadataAndSerializationContext.Default.PersonStruct.SerializeHandler);
+            Assert.NotNull(MetadataAndSerializationContext.Default.TypeWithValidationAttributes.SerializeHandler);
+            Assert.NotNull(MetadataAndSerializationContext.Default.TypeWithDerivedAttribute.SerializeHandler);
         }
     }
 }

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/MetadataContextTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/MetadataContextTests.cs
@@ -40,6 +40,8 @@ namespace System.Text.Json.SourceGeneration.Tests
     [JsonSerializable(typeof(ClassWithBadCustomConverter), GenerationMode = JsonSourceGenerationMode.Metadata)]
     [JsonSerializable(typeof(StructWithBadCustomConverter), GenerationMode = JsonSourceGenerationMode.Metadata)]
     [JsonSerializable(typeof(PersonStruct?), GenerationMode = JsonSourceGenerationMode.Metadata)]
+    [JsonSerializable(typeof(TypeWithValidationAttributes), GenerationMode = JsonSourceGenerationMode.Metadata)]
+    [JsonSerializable(typeof(TypeWithDerivedAttribute), GenerationMode = JsonSourceGenerationMode.Metadata)]
     internal partial class MetadataWithPerTypeAttributeContext : JsonSerializerContext, ITestContext
     {
         public JsonSourceGenerationMode JsonSourceGenerationMode => JsonSourceGenerationMode.Metadata;
@@ -86,6 +88,8 @@ namespace System.Text.Json.SourceGeneration.Tests
             Assert.Throws<InvalidOperationException>(() => MetadataWithPerTypeAttributeContext.Default.StructWithBadCustomConverter.SerializeHandler);
             Assert.Null(MetadataWithPerTypeAttributeContext.Default.NullablePersonStruct.SerializeHandler);
             Assert.Null(MetadataWithPerTypeAttributeContext.Default.PersonStruct.SerializeHandler);
+            Assert.Null(MetadataWithPerTypeAttributeContext.Default.TypeWithValidationAttributes.SerializeHandler);
+            Assert.Null(MetadataWithPerTypeAttributeContext.Default.TypeWithDerivedAttribute.SerializeHandler);
         }
     }
 
@@ -124,6 +128,8 @@ namespace System.Text.Json.SourceGeneration.Tests
     [JsonSerializable(typeof(ClassWithBadCustomConverter))]
     [JsonSerializable(typeof(StructWithBadCustomConverter))]
     [JsonSerializable(typeof(PersonStruct?))]
+    [JsonSerializable(typeof(TypeWithValidationAttributes))]
+    [JsonSerializable(typeof(TypeWithDerivedAttribute))]
     internal partial class MetadataContext : JsonSerializerContext, ITestContext
     {
         public JsonSourceGenerationMode JsonSourceGenerationMode => JsonSourceGenerationMode.Metadata;
@@ -193,6 +199,8 @@ namespace System.Text.Json.SourceGeneration.Tests
             Assert.Throws<InvalidOperationException>(() => MetadataContext.Default.StructWithBadCustomConverter.SerializeHandler);
             Assert.Null(MetadataContext.Default.NullablePersonStruct.SerializeHandler);
             Assert.Null(MetadataContext.Default.PersonStruct.SerializeHandler);
+            Assert.Null(MetadataContext.Default.TypeWithValidationAttributes.SerializeHandler);
+            Assert.Null(MetadataContext.Default.TypeWithDerivedAttribute.SerializeHandler);
         }
 
         [Fact]

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/MixedModeContextTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/MixedModeContextTests.cs
@@ -41,6 +41,8 @@ namespace System.Text.Json.SourceGeneration.Tests
     [JsonSerializable(typeof(ClassWithBadCustomConverter), GenerationMode = JsonSourceGenerationMode.Metadata | JsonSourceGenerationMode.Serialization)]
     [JsonSerializable(typeof(StructWithBadCustomConverter), GenerationMode = JsonSourceGenerationMode.Metadata | JsonSourceGenerationMode.Serialization)]
     [JsonSerializable(typeof(PersonStruct?), GenerationMode = JsonSourceGenerationMode.Metadata | JsonSourceGenerationMode.Serialization)]
+    [JsonSerializable(typeof(TypeWithValidationAttributes), GenerationMode = JsonSourceGenerationMode.Metadata | JsonSourceGenerationMode.Serialization)]
+    [JsonSerializable(typeof(TypeWithDerivedAttribute), GenerationMode = JsonSourceGenerationMode.Metadata | JsonSourceGenerationMode.Serialization)]
     internal partial class MixedModeContext : JsonSerializerContext, ITestContext
     {
         public JsonSourceGenerationMode JsonSourceGenerationMode => JsonSourceGenerationMode.Metadata | JsonSourceGenerationMode.Serialization;
@@ -88,6 +90,8 @@ namespace System.Text.Json.SourceGeneration.Tests
             Assert.Throws<InvalidOperationException>(() => MixedModeContext.Default.StructWithBadCustomConverter.SerializeHandler);
             Assert.Null(MixedModeContext.Default.NullablePersonStruct.SerializeHandler);
             Assert.NotNull(MixedModeContext.Default.PersonStruct.SerializeHandler);
+            Assert.NotNull(MixedModeContext.Default.TypeWithValidationAttributes.SerializeHandler);
+            Assert.NotNull(MixedModeContext.Default.TypeWithDerivedAttribute.SerializeHandler);
         }
 
         [Fact]

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs
@@ -886,10 +886,10 @@ namespace System.Text.Json.SourceGeneration.Tests
         [Fact]
         public void TypeWithValidationAttributes()
         {
-            var instance = new TypeWithValidationAttributes { Name = "Test Name" };
+            var instance = new TypeWithValidationAttributes { Name = "Test Name", Email = "email@test.com" };
 
             string json = JsonSerializer.Serialize(instance, DefaultContext.TypeWithValidationAttributes);
-            JsonTestHelper.AssertJsonEqual(@"{""Name"":""Test Name""}", json);
+            JsonTestHelper.AssertJsonEqual(@"{""Name"":""Test Name"",""Email"":""email@test.com""}", json);
             if (DefaultContext.JsonSourceGenerationMode == JsonSourceGenerationMode.Serialization)
             {
                 // Deserialization not supported in fast path serialization only mode
@@ -899,6 +899,7 @@ namespace System.Text.Json.SourceGeneration.Tests
             {
                 instance = JsonSerializer.Deserialize(json, DefaultContext.TypeWithValidationAttributes);
                 Assert.Equal("Test Name", instance.Name);
+                Assert.Equal("email@test.com", instance.Email);
             }
         }
 

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs
@@ -882,5 +882,43 @@ namespace System.Text.Json.SourceGeneration.Tests
             Assert.Equal("Jane", person.Value.FirstName);
             Assert.Equal("Doe", person.Value.LastName);
         }
+
+        [Fact]
+        public void TypeWithValidationAttributes()
+        {
+            var instance = new TypeWithValidationAttributes { Name = "Test Name" };
+
+            string json = JsonSerializer.Serialize(instance, DefaultContext.TypeWithValidationAttributes);
+            JsonTestHelper.AssertJsonEqual(@"{""Name"":""Test Name""}", json);
+            if (DefaultContext.JsonSourceGenerationMode == JsonSourceGenerationMode.Serialization)
+            {
+                // Deserialization not supported in fast path serialization only mode
+                Assert.Throws<InvalidOperationException>(() => JsonSerializer.Deserialize(json, DefaultContext.TypeWithValidationAttributes));
+            }
+            else
+            {
+                instance = JsonSerializer.Deserialize(json, DefaultContext.TypeWithValidationAttributes);
+                Assert.Equal("Test Name", instance.Name);
+            }
+        }
+
+        [Fact]
+        public void TypeWithDerivedAttribute()
+        {
+            var instance = new TypeWithDerivedAttribute();
+
+            string json = JsonSerializer.Serialize(instance, DefaultContext.TypeWithDerivedAttribute);
+            JsonTestHelper.AssertJsonEqual(@"{}", json);
+            if (DefaultContext.JsonSourceGenerationMode == JsonSourceGenerationMode.Serialization)
+            {
+                // Deserialization not supported in fast path serialization only mode
+                Assert.Throws<InvalidOperationException>(() => JsonSerializer.Deserialize(json, DefaultContext.TypeWithDerivedAttribute));
+            }
+            else
+            {
+                instance = JsonSerializer.Deserialize(json, DefaultContext.TypeWithDerivedAttribute);
+                Assert.NotNull(instance);
+            }
+        }
     }
 }

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/SerializationContextTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/SerializationContextTests.cs
@@ -41,6 +41,8 @@ namespace System.Text.Json.SourceGeneration.Tests
     [JsonSerializable(typeof(ClassWithBadCustomConverter))]
     [JsonSerializable(typeof(StructWithBadCustomConverter))]
     [JsonSerializable(typeof(PersonStruct?))]
+    [JsonSerializable(typeof(TypeWithValidationAttributes))]
+    [JsonSerializable(typeof(TypeWithDerivedAttribute))]
     internal partial class SerializationContext : JsonSerializerContext, ITestContext
     {
         public JsonSourceGenerationMode JsonSourceGenerationMode => JsonSourceGenerationMode.Serialization;
@@ -80,6 +82,8 @@ namespace System.Text.Json.SourceGeneration.Tests
     [JsonSerializable(typeof(ClassWithBadCustomConverter), GenerationMode = JsonSourceGenerationMode.Serialization)]
     [JsonSerializable(typeof(StructWithBadCustomConverter), GenerationMode = JsonSourceGenerationMode.Serialization)]
     [JsonSerializable(typeof(PersonStruct?), GenerationMode = JsonSourceGenerationMode.Serialization)]
+    [JsonSerializable(typeof(TypeWithValidationAttributes), GenerationMode = JsonSourceGenerationMode.Serialization)]
+    [JsonSerializable(typeof(TypeWithDerivedAttribute), GenerationMode = JsonSourceGenerationMode.Serialization)]
     internal partial class SerializationWithPerTypeAttributeContext : JsonSerializerContext, ITestContext
     {
         public JsonSourceGenerationMode JsonSourceGenerationMode => JsonSourceGenerationMode.Serialization;
@@ -120,6 +124,8 @@ namespace System.Text.Json.SourceGeneration.Tests
     [JsonSerializable(typeof(ClassWithBadCustomConverter), GenerationMode = JsonSourceGenerationMode.Serialization)]
     [JsonSerializable(typeof(StructWithBadCustomConverter), GenerationMode = JsonSourceGenerationMode.Serialization)]
     [JsonSerializable(typeof(PersonStruct?), GenerationMode = JsonSourceGenerationMode.Serialization)]
+    [JsonSerializable(typeof(TypeWithValidationAttributes), GenerationMode = JsonSourceGenerationMode.Serialization)]
+    [JsonSerializable(typeof(TypeWithDerivedAttribute), GenerationMode = JsonSourceGenerationMode.Serialization)]
     internal partial class SerializationContextWithCamelCase : JsonSerializerContext, ITestContext
     {
         public JsonSourceGenerationMode JsonSourceGenerationMode => JsonSourceGenerationMode.Serialization;
@@ -169,6 +175,8 @@ namespace System.Text.Json.SourceGeneration.Tests
             Assert.Throws<InvalidOperationException>(() => SerializationContext.Default.StructWithBadCustomConverter.SerializeHandler);
             Assert.Null(SerializationContext.Default.NullablePersonStruct.SerializeHandler);
             Assert.NotNull(SerializationContext.Default.PersonStruct.SerializeHandler);
+            Assert.NotNull(SerializationContext.Default.TypeWithValidationAttributes.SerializeHandler);
+            Assert.NotNull(SerializationContext.Default.TypeWithDerivedAttribute.SerializeHandler);
         }
 
         [Fact]
@@ -492,6 +500,8 @@ namespace System.Text.Json.SourceGeneration.Tests
             Assert.Throws<InvalidOperationException>(() => SerializationWithPerTypeAttributeContext.Default.StructWithBadCustomConverter.SerializeHandler);
             Assert.Null(SerializationWithPerTypeAttributeContext.Default.NullablePersonStruct.SerializeHandler);
             Assert.NotNull(SerializationWithPerTypeAttributeContext.Default.PersonStruct.SerializeHandler);
+            Assert.NotNull(SerializationWithPerTypeAttributeContext.Default.TypeWithValidationAttributes.SerializeHandler);
+            Assert.NotNull(SerializationWithPerTypeAttributeContext.Default.TypeWithDerivedAttribute.SerializeHandler);
         }
     }
 }

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/TestClasses.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/TestClasses.cs
@@ -164,6 +164,9 @@ namespace System.Text.Json.SourceGeneration.Tests
         [ComponentModel.DataAnnotations.Required(ErrorMessage = "Name is required")]
         [ComponentModel.DataAnnotations.StringLength(100, ErrorMessage = "Name must not be longer than 100 characters")]
         public string Name { get; set; }
+
+        [ComponentModel.DataAnnotations.Required]
+        public string Email { get; set; }
     }
 
     public class BaseAttribute : Attribute

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/TestClasses.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/TestClasses.cs
@@ -158,4 +158,23 @@ namespace System.Text.Json.SourceGeneration.Tests
         public string FirstName { get; set; }
         public string LastName { get; set; }
     }
+
+    public class TypeWithValidationAttributes
+    {
+        [ComponentModel.DataAnnotations.Required(ErrorMessage = "Name is required")]
+        [ComponentModel.DataAnnotations.StringLength(100, ErrorMessage = "Name must not be longer than 100 characters")]
+        public string Name { get; set; }
+    }
+
+    public class BaseAttribute : Attribute
+    {
+        public string TestProperty { get; set; }
+    }
+
+    public class DerivedAttribute : BaseAttribute
+    { }
+
+    [Derived(TestProperty = "Test")]
+    public class TypeWithDerivedAttribute
+    { }
 }

